tests/integration/test_active_learning.py

```diff
@@ -0,0 +1,180 @@
+    criterion = 0.01 * test_range
+
+    # we expect a model with initial data to fail the criterion
+    initial_model = build_model(initial_data)
+    initial_model.optimize(initial_data)
+    initial_predicted_means, _ = initial_model.model.predict_f(test_query_points)  # type: ignore
+    initial_accuracy = tf.reduce_max(tf.abs(initial_predicted_means - test_data.observations))
+
+    assert not initial_accuracy < criterion
+
+    # after active learning the model should be much more accurate
+    model = build_model(initial_data)
+    final_model = (
+        BayesianOptimizer(observer, search_space)
+        .optimize(num_steps, initial_data, model, acquisition_rule)
+        .try_get_final_model()
+    )
+    final_predicted_means, _ = final_model.model.predict_f(test_query_points)  # type: ignore
+    final_accuracy = tf.reduce_max(tf.abs(final_predicted_means - test_data.observations))
+
+    assert initial_accuracy < final_accuracy
+    assert final_accuracy < criterion
+
+
+@random_seed
+@pytest.mark.slow
+@pytest.mark.parametrize(
+    "num_steps, acquisition_rule, threshold",
+    [
+        (50, EfficientGlobalOptimization(ExpectedFeasibility(80, delta=1)), 80),
+        (50, EfficientGlobalOptimization(ExpectedFeasibility(80, delta=2)), 80),
+        (50, EfficientGlobalOptimization(ExpectedFeasibility(20, delta=1)), 20),
+    ],
+)
+def test_optimizer_learns_feasibility_set_of_thresholded_branin_function(
+    num_steps: int, acquisition_rule: AcquisitionRule[TensorType, SearchSpace], threshold: int
+) -> None:
+    """
+    Ensure that the feasible set is sufficiently well learned, such that the final model
+    classifies with great degree of certainty whether points in the search space are in
+    in the feasible set or not.
+    """
+
+    search_space = BRANIN_SEARCH_SPACE
+
+    def build_model(data: Dataset) -> TrainableProbabilisticModel:
+        variance = tf.math.reduce_variance(data.observations)
+        kernel = gpflow.kernels.Matern52(variance=variance, lengthscales=[0.2, 0.2])
+        prior_scale = tf.cast(1.0, dtype=tf.float64)
+        kernel.variance.prior = tfp.distributions.LogNormal(
+            tf.cast(-2.0, dtype=tf.float64), prior_scale
+        )
+        kernel.lengthscales.prior = tfp.distributions.LogNormal(
+            tf.math.log(kernel.lengthscales), prior_scale
+        )
+        gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+        gpflow.set_trainable(gpr.likelihood, False)
+
+        return GaussianProcessRegression(gpr)
+
+    num_initial_points = 6
+    initial_query_points = search_space.sample_halton(num_initial_points)
+    observer = mk_observer(branin)
+    initial_data = observer(initial_query_points)
+
+    # we set a performance criterion at 0.001 probability of required precision per point
+    n_test = 10000 * search_space.dimension
+    test_query_points = search_space.sample_sobol(n_test)
```

Comment:
I think we are more interested in model accuracy around the threshold (or perhaps the whole feasible region?), instead of getting test points from the whole search space, we should then collect test points around the boundary - we could do something like:
```
test_query_points = search_space.sample(100000)
test_data = observer(test_query_points)

mask = tf.reduce_all(tf.concat([test_data.observations > 80 - 5, test_data.observations < 80 + 5], axis = 1), axis=1)
threshold_test_points = tf.boolean_mask(test_query_points, mask)
assert threshold_test_points.shape[0] > 1000
```

Reply:
I'm not sure I 100% agree, we also want our GP to be good enough to predict anywhere which side it's on with high probability. I think the random sample is fine.

We could also have 2 criteria, one 'global' as it is and one 'local' as you describe and we could check both. Not sure it's worth the extra effort

Reply:
I went with the 2 criteria version...